Navigation component: Add fade-in animation to badge

david-szabo97 opened the pull request:
Fixes #24950 

## Description
Once the value of `badge` is truthy, the badge fades in. The badge also fades-in on mount.

## How has this been tested?
`yarn run storybook:dev`

## Screenshots
![](https://user-images.githubusercontent.com/2256104/93452457-bd410200-f8d8-11ea-8ca7-dd746100ea5b.gif)

## Types of changes
Enhancement 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

david-szabo97 committed:
Show spinner if isFetchingBadge set

david-szabo97 committed:
Remove spinner margin

david-szabo97 committed:
Add example

david-szabo97 commented:
Up for discussion. Happy to add tests and documentation once we agreed on all the technical details.

shaunandrews commented:
I think it would be better to show _nothing_ rather than this spinner; It feels too distracting for information that isn't really that important. Instead, I'd expect the badge to transition (probably just fade in) into place when the data is loaded. Otherwise, I worry we'll end up with half-a-dozen spinners in the sidebar.

david-szabo97 committed:
Remove spinner

david-szabo97 committed:
Add fade-in animation for badge

david-szabo97 committed:
Update story

david-szabo97 commented:
@shaunandrews Agree. Multiple spinners would be annoying that's for sure.
Updated the PR and replaced spinner with a fade-in animation. How do you like it?

Right now it also animates the badge if it has a value at the time of the mount. Probably we should get rid of that. 

![](https://user-images.githubusercontent.com/2256104/93452457-bd410200-f8d8-11ea-8ca7-dd746100ea5b.gif)

Copons commented on packages/components/src/navigation/styles/navigation-styles.js:
```diff
@@ -95,6 +95,20 @@ export const ItemBadgeUI = styled.span`
+		to {
+			opacity: 1;
+		}
+	}
```
We have a mixin for fade-ins: [`edit-post__fade-in-animation`](https://github.com/WordPress/gutenberg/blob/50955c1f8ee632a34a8faea2b22bf62fbaefae5a/packages/base-styles/_animations.scss).
The name is counter-intuitive, but I've see it used for components as well.

I'm not familiar enough with styled components to know if it's possible to import a SASS mixin.
If we can't, I think we should at least use the allow folks to [opt out of animations](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) with [`reduce-motion`](https://github.com/WordPress/gutenberg/blob/50955c1f8ee632a34a8faea2b22bf62fbaefae5a/packages/base-styles/_mixins.scss#L214-L239).
We even have a handy [JS utility](https://github.com/WordPress/gutenberg/blob/50955c1f8ee632a34a8faea2b22bf62fbaefae5a/packages/components/src/utils/reduce-motion.js) for that already! ✨ 

I think it should be used like this:
```js
const ItemBadgeUI = styled.span`
  animation: fade-in 250ms ease-out;
  @keyframes fade-in { /* ... */ }

  ${ reduceMotion( 'animation' ) };
`;
```

david-szabo97 replied:
We can't use SASS in Styled :( How about creating a styles file? We already have a class for the badge: 
```scss
.components-navigation__item-badge {
	@include edit-post__fade-in-animation;
}
```

Later on we should recreate this mixin in JS though so we can use it with Styled.

Copons replied:
I think we should follow Gutenberg's own direction, which unfortunately I don't know. 😅 

If we're staying with SASS, then sure let's add a `style.scss` file!
If we're moving to styled components, then let's go with a fade-in JS mixin!

Either way, I think in this specific case it's fine to keep what you did, and just add the `${ reduceMotion( 'animation' ) }`.

Copons commented:
This is ok from a code standpoint (`reduce-motion` suggestion notwhitstanding), and the animation is subtle enough that it's probably not a big deal to keep it on mount.

I don't have any strong opinion on this design-wise, so I'll let @shaunandrews have the final say.

david-szabo97 committed:
Remove unused spinner style

david-szabo97 committed:
Add reduce motion

david-szabo97 committed:
Rename example item's title

david-szabo97 committed:
Merge branch 'master' into try/navigation-badge-spinner

david-szabo97 committed:
Merge branch 'master' into try/navigation-badge-spinner

david-szabo97 committed:
Fix spacing

Addison-Stavlo commented on packages/components/src/navigation/stories/index.js:
```diff
@@ -30,6 +30,12 @@ function Example() {
 
+	const [ delayedBadge, setDelayedBadge ] = useState();
+	useEffect( () => {
+		const timeout = setTimeout( () => setDelayedBadge( 2 ), 1500 );
+		return () => clearInterval( timeout );
```
Should this be `clearTimeout` instead of `clearInterval` ?  I don't think using `clearInterval` to clear a timeout will be guaranteed to work in all cases.

david-szabo97 replied:
Good catch! My bad 😬 I was testing with setInterval and later on replaced it with setTimeout but forgot the clear* 😞 I'll open a PR to fix it.
